bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/handler/VerisureAlarmThingHandler.java

```diff
@@ -0,0 +1,208 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.verisure.internal.handler;
+
+import static org.openhab.binding.verisure.internal.VerisureBindingConstants.*;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.verisure.internal.VerisureSession;
+import org.openhab.binding.verisure.internal.model.VerisureAlarms;
+import org.openhab.binding.verisure.internal.model.VerisureAlarms.ArmState;
+import org.openhab.binding.verisure.internal.model.VerisureThing;
+
+/**
+ * Handler for the Alarm Device thing type that Verisure provides.
+ *
+ * @author Jan Gustafsson - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class VerisureAlarmThingHandler extends VerisureThingHandler {
```

Comment:
I forgot to mention the last step of my previous request. Since VerisureThingHandler requires a generic parameter, you now need to specify that parameter for each of the subclasses. Once you specify the VerisureThing type, your `update` method will also update its type accordingly.
```suggestion
public class VerisureAlarmThingHandler extends VerisureThingHandler<VerisureAlarms> {
```

Reply:
OK, does that mean that the update method in VerisureThingHandler should look like this:

```
public synchronized void update(T thing) {
        ChannelUID cuid = new ChannelUID(getThing().getUID(), CHANNEL_INSTALLATION_ID);
        BigDecimal siteId = thing.getSiteId();
        updateState(cuid, new DecimalType(siteId.longValue()));
        cuid = new ChannelUID(getThing().getUID(), CHANNEL_INSTALLATION_NAME);
        updateState(cuid, new StringType(thing.getSiteName()));
 }
```
I then have to adapt the other method calling update on VTH, since that does not know the class type of the thing when calling update:

```
@Override
    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
        logger.debug("bridgeStatusChanged bridgeStatusInfo: {}", bridgeStatusInfo);
        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE) {
            VerisureSession session = getSession();
            if (session != null) {
                String deviceId = config.getDeviceId();
                VerisureThing thing = session.getVerisureThing(deviceId);
                if (thing != null) {
                    update(thing);
                } else {
                    logger.warn("Thing is null!");
                }
                session.registerDeviceStatusListener(this);
            }
        }
        super.bridgeStatusChanged(bridgeStatusInfo);
}
```

What is the best way of adapting that call?

Reply:
Yeah, my idea was to make the `update` method use a generic parameter as well, but you brought up an issue I didn't foresee so made another comment below explaining a way to fix it...hopefully. If the the changes get too complicated I'll checkout from your branch and see if I can resolve them. Thank you for putting up with this request so far. You have been doing a great job.